org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/BaseInitHandler.java

```diff
@@ -57,19 +62,59 @@ public BaseInitHandler(ProjectsManager projectsManager, PreferenceManager prefer
 		this.projectsManager = projectsManager;
 	}
 
-	@SuppressWarnings("unchecked")
 	public InitializeResult initialize(InitializeParams param) {
 		logInfo("Initializing Java Language Server " + JavaLanguageServerPlugin.getVersion());
 		InitializeResult result = new InitializeResult();
 		handleInitializationOptions(param);
 		registerCapabilities(result);
+		configureContentTypes();
 
 		// At the end of the InitHandler, trigger a job to import the workspace. This is used to ensure ServiceStatus notification
 		// is not sent before the initialize response. See the bug https://github.com/redhat-developer/vscode-java/issues/1056
 		triggerInitialization(preferenceManager.getPreferences().getRootPaths());
 		return result;
 	}
 
+	private void configureContentTypes() {
+		if (preferenceManager.getPreferences().getFilesAssociations() != null) {
+			IContentType javaSourceContentType = Platform.getContentTypeManager().getContentType(JavaCore.JAVA_SOURCE_CONTENT_TYPE);
+			if (javaSourceContentType != null) {
+				List<String> toRemove = new ArrayList<>();
+				String[] specs = javaSourceContentType.getFileSpecs(IContentType.FILE_EXTENSION_SPEC);
+				for (String spec : specs) {
+					if (!SuffixConstants.EXTENSION_java.equals(spec)) {
+						toRemove.add(spec);
+					}
+				}
+				List<String> toAdd = new ArrayList<>();
+				for (String spec : preferenceManager.getPreferences().getFilesAssociations()) {
+					if (toRemove.contains(spec)) {
+						toRemove.remove(spec);
+					} else {
+						toAdd.add(spec);
+					}
+				}
```

Comment:
Just a comment. Although I think it's possible to restate this entire code in terms of using `List.removeAll` on the opposite list (`toRemove=oldSpecs, toAdd=newSpecs, toRemove.removeAll(newSpecs), toAdd.removeAll(oldSpecs))`, and just not touching the "java" spec, I actually think this approach is much more readable so this is fine with me.